### PR TITLE
Fix new ensemble sometimes failed problem.

### LINKED ISSUE #3722: RackAware placement policy can not ensure new ensemble selection succeed when one rack goes down.
BUG REPORT

Describe the bug

r1 -> bk1, bk2
r2 -> bk3
r3 -> bk4
enable EnforceMinNumRacksPerWriteQuorum and set minNumRacksPerWriteQuorumConfValue=2

When bk3 or bk4 is quarantined, the new ensemble selection should succeed, because it fulfills min rack is 2.
However, the new ensemble selection result some times will fail.

Run the following unit test in TestRackawareEnsemblePlacementPolicy.java can reproduce the bug.

```
final int minNumRacksPerWriteQuorumConfValue = 2;
conf.setMinNumRacksPerWriteQuorum(minNumRacksPerWriteQuorumConfValue);
conf.setEnforceMinNumRacksPerWriteQuorum(true);

@Test
    public void testNewEnsemblePolicyWithMultipleRacksV2() throws Exception {
        BookieSocketAddress addr1 = new BookieSocketAddress("127.0.0.1", 3181);
        BookieSocketAddress addr2 = new BookieSocketAddress("127.0.0.2", 3181);
        BookieSocketAddress addr3 = new BookieSocketAddress("127.0.0.3", 3181);
        BookieSocketAddress addr4 = new BookieSocketAddress("127.0.0.4", 3181);

        // update dns mapping
        StaticDNSResolver.addNodeToRack(addr1.getHostName(), "/default-region/r1");
        StaticDNSResolver.addNodeToRack(addr2.getHostName(), "/default-region/r1");
        StaticDNSResolver.addNodeToRack(addr3.getHostName(), "/default-region/r2");
        StaticDNSResolver.addNodeToRack(addr4.getHostName(), "/default-region/r3");
        // Update cluster
        Set<BookieId> addrs = new HashSet<BookieId>();
        addrs.add(addr1.toBookieId());
        addrs.add(addr2.toBookieId());
        addrs.add(addr3.toBookieId());
        addrs.add(addr4.toBookieId());
        repp.onClusterChanged(addrs, new HashSet<BookieId>());

        try {
            int ensembleSize = 3;
            int writeQuorumSize = 3;
            int ackQuorumSize = 2;

            Set<BookieId> excludeBookies = new HashSet<>();
            excludeBookies.add(addr4.toBookieId());
            //excludeBookies.add(addr3.toBookieId());

            for (int i = 0; i < 50; ++i) {
                EnsemblePlacementPolicy.PlacementResult<List<BookieId>> ensembleResponse =
                    repp.newEnsemble(ensembleSize, writeQuorumSize,
                        ackQuorumSize, null, excludeBookies);
                List<BookieId> ensemble = ensembleResponse.getResult();
            }
        } catch (Exception e ) {
            fail("Can not new ensemble selection succeed");
        }
    }
```

### PR DESCRIPTION
**Pull request by horizonzy**
Descriptions of the changes in this PR:

Fixes #3722

### CODE REVIEW

**Review by hangc0276 — COMMENTED** [review]
Please add an explanation about the reason why it failed to select ensembles out sometimes.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 1]
```diff
@@ -678,6 +728,38 @@ protected BookieNode selectRandomFromRack(String netPath, Set<Node> excludeBooki
...
+    /**
+     * Sort the node list by the network location.
+     * The higher the frequency of the network location, it will on the list head.
+     */
+    private List<Node> sortByCommonRack(List<Node> nodeLists) {
+        Map<String, Integer> map = new HashMap<>();
+        for (Node node : nodeLists) {
+            map.put(node.getNetworkLocation(), map.getOrDefault(node.getNetworkLocation(), 0) + 1);
+        }
+        List<Object[]> list = new ArrayList<>();
```
Comment (StevenLuMT) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java:740]: Do you need to set the length of the list in advance?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 2]
```diff
@@ -444,6 +440,33 @@ protected PlacementResult<List<BookieId>> newEnsembleInternal(
...
+        for (int i = 0; i < ensembleSize; i++) {
+            String curRack;
+            if (null == prevNode) {
+                if ((null == localNode) || defaultRack.equals(localNode.getNetworkLocation())) {
+                    curRack = NodeBase.ROOT;
+                } else {
+                    curRack = localNode.getNetworkLocation();
+                }
+            } else {
+                curRack = "~" + prevNode.getNetworkLocation();
```
Comment (StevenLuMT) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java:457]: "~" is the magic string

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ITopologyAwareEnsemblePlacementPolicy.java [hunk 3]
```diff
@@ -122,6 +122,29 @@ T selectFromNetworkLocation(String networkLoc,
...
+     * @param predicate
+     *          predicate to apply
+     * @param ensemble
+     *          ensemble
+     * @param fallbackToRandom
+     *          fallbackToRandom
+     * @return the selected bookie.
+     * @throws BKException.BKNotEnoughBookiesException
+     */
+    T selectFromNetworkLocationWithPickCommonRackFirst(String networkLoc,
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ITopologyAwareEnsemblePlacementPolicy.java:141]: Need a default implementation.

**Review by zymap — COMMENTED** [review]
Overall looks good to me.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java (hunk 1 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java:740]: The size is not unpredictable

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ITopologyAwareEnsemblePlacementPolicy.java (hunk 3 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ITopologyAwareEnsemblePlacementPolicy.java:141]: Followed the method `selectFromNetworkLocation`; it's not a default method. In the `ZoneawareEnsemblePlacementPolicyImpl`, it implements the method and throw exception.
https://github.com/apache/bookkeeper/blob/405e72acf42bb1104296447ea8840d805094c787/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ZoneawareEnsemblePlacementPolicyImpl.java#L303-L309

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java (hunk 2 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java:457]: Yes, I will extract it.

**Comment by horizonzy** [comment]
> Please add an explanation about the reason why it failed to select ensembles out sometimes.

See #3722

**Comment by horizonzy** [comment]
rerun failure checks

**Comment by horizonzy** [comment]
This fix can't cover the extreme case: 
8 rack1, 1 rack2, 1 rack-3.
```
E:10 
WQ:10 
EQ:2
enforceMinNumRacksPerWriteQuorum:true
minNumRacksPerWriteQuorum:2
```
It still can't pick newEnsemble.

**Comment by horizonzy** [comment]
New PR: #3725

**Pull request closed without merging**